### PR TITLE
[WIP] Feature: Warning system for #47

### PR DESCRIPTION
**Pull request by Matslom**
#47 
 - [ ] Need some js
- [ ] Make better design
- [X] Warn a user
 - [X] Warn a user for a particular piece of content (post, topic etc)
 - [X] Configurable presets for warning points and reasons
 - [X] Warn using a preset, or custom reason and points
- [X] Revoke warn
- [x] Acknowledge with warn

### CODE REVIEW

File: app/Database/Repositories/Eloquent/WarningsRepository.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        $warning = $this->warningModel->create($details);
+        return $warning;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findForUser($userId)
+    {
+        return $this->warningModel->where('user_id', $userId)->with(['issuedBy'])->orderBy('created_at', 'desc')->get()->groupBy('expired');
```
Comment (euantorano) [inline app/Database/Repositories/Eloquent/WarningsRepository.php]: Could this be broken across a couple of lines please, just to make it easier to read?

File: app/Database/Repositories/Eloquent/WarningsRepository.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function find($warnId)
+    {
+        return $this->warningModel->find($warnId);
+    }
+
```
Comment (euantorano) [inline app/Database/Repositories/Eloquent/WarningsRepository.php]: And another empty line 😄

File: app/Database/Repositories/WarningsRepositoryInterface.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    /**
+     * Get all users.
+     *
+     * @param string $sortBy
+     * @param string $sortDir
+     * @param int $perPage
+     *
+     * @return mixed
+     */
+    //public function allForUser($user_id);
```
Comment (euantorano) [inline app/Database/Repositories/WarningsRepositoryInterface.php]: Is this function going to be implemented, or can it be removed from the interface?

File: app/Http/Controllers/Admin/Users/WarningsController.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        $this->warningTypesRepository->create([
+            'reason'              => $request->input('reason'),
+            'points'              => $request->input('points'),
+            'expiration_type'     => $request->input('type'),
+            'expiration_multiple' => $request->input('multiple'),
+            'must_acknowledge'    => $request->input('must_acknowledge'),
+        ]);
+
+        return redirect()->route('admin.warnings.warning_types')->withSuccess(trans('admin::warnings.warning_type_create_success'));
```
Comment (euantorano) [inline app/Http/Controllers/Admin/Users/WarningsController.php]: Could this be split across multiple lines too please?

File: app/Http/Controllers/Admin/Users/WarningsController.php [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+        $this->warningTypesRepository->edit($warningType, [
+            'reason'              => $request->input('reason'),
+            'points'              => $request->input('points'),
+            'expiration_type'     => $request->input('type'),
+            'expiration_multiple' => $request->input('multiple'),
+            'must_acknowledge'    => $request->input('must_acknowledge'),
+        ]);
+
+        return redirect()->route('admin.warnings.warning_types')->withSuccess(trans('admin::warnings.warning_type_edit_success'));
```
Comment (euantorano) [inline app/Http/Controllers/Admin/Users/WarningsController.php]: And another long line.

File: app/Http/Controllers/WarningsController.php [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        $warningContent = $this->WarningsManager->getWarningContentClass($contentType);
+        $content = $warningContent->getWarningContent($contentId);
+        if ($user->id != $content['user_id'])
+            throw new UserNotBelongsToThisContentException($user->name);
+
+        $previewContent = $warningContent->getWarningPreviewView($content['content']);
+
+        $warningTypes = $this->warningTypesRepository->all();
+
+        return view('warnings.warn_user', compact('user', 'contentType', 'contentId', 'warningTypes', 'previewContent'));
```
Comment (euantorano) [inline app/Http/Controllers/WarningsController.php]: Another long line that could be split across multiple lines please.

**Comment by euantorano** [comment]
Also some code style changes needed it seems.

Thanks for the hard work so far though - this looks really well done!

**Comment by euantorano** [comment]
Looks good to me - I remember you saying you weren't too good at JS, so do you want to leave that and either I can add it later or somebody else can?

**Comment by Matslom** [comment]
My js is very messy + I did not do anything with typescript previous. It will be nice if someone can do this.

**Comment by euantorano** [comment]
No problem, I can do that :)

**Comment by premiumcuyuz** [comment]
good work :) @euantorano @Matslom We are waiting for more 👍

**Comment by 036** [comment]
Looks good @Matslom good job.

**Comment by euantorano** [comment]
Looking very good, thanks!

**Pull request merged**